catalogue/config/util.py

```diff
@@ -0,0 +1,59 @@
+import contextlib
+import functools
+import shutil
+import tempfile
+from pathlib import Path
+from typing import TypeVar, Callable, Any, Iterator
+
+try:
+    from typing import Protocol
+except ImportError:
+    # Ignoring type for mypy to avoid "Incompatible import" error (https://github.com/python/mypy/issues/4427).
+    from typing_extensions import Protocol  # type: ignore
```

Comment:
The original code has 
```
if sys.version_info < (3, 8):
    from typing_extensions import Protocol
else:
    from typing import Protocol
```
which is slightly better IMO as we try not to rely on exception raising to handle flow.